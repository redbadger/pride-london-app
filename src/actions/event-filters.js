// @flow
import type { DateRange, Time } from "../data/date-time";
import type { StandardAction } from "./";

type EventFiltersActionType =
  | "STAGE_EVENT_FILTERS"
  | "COMMIT_EVENT_FILTERS"
  | "CLEAR_STAGED_EVENT_FILTERS";
type EventFiltersPayload = {
  date?: ?DateRange,
  time?: Set<Time>
};

export type EventFiltersAction = StandardAction<
  EventFiltersActionType,
  EventFiltersPayload
>;

export const stageEventFilters = (updates: EventFiltersPayload) => (
  dispatch: Dispatch<EventFiltersAction>
) => {
  dispatch({
    type: "STAGE_EVENT_FILTERS",
    payload: updates
  });
};

export const commitEventFilters = () => (
  dispatch: Dispatch<EventFiltersAction>
) => dispatch({ type: "COMMIT_EVENT_FILTERS" });

export const clearStagedEventFilters = () => (
  dispatch: Dispatch<EventFiltersAction>
) => dispatch({ type: "CLEAR_STAGED_EVENT_FILTERS" });
