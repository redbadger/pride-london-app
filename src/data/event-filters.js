// @flow
import type { DateRange, Time } from "./date-time";
import type { EventCategoryName } from "./event";

export type Area = "Central" | "East" | "North" | "South" | "West";
export type Price = "free";
export type Audience = "Families" | "Youth" | "16+" | "18+";
export type VenueDetail = "Gender neutral toilets" | "Outdoors" | "Indoors";
export type AccessibilityOption =
  | "Step free access"
  | "Wheelchair accessibile"
  | "Accessible Toilets"
  | "BSL Interpreter"
  | "Hearing loop installed"
  | "Complimentary ticket for personal assistant"
  | "Service animals welcome"
  | "Accessible parking spaces at venue";

export type FilterCollection = {
  date: ?DateRange,
  timeOfDay: Set<Time>,
  categories: Set<EventCategoryName>,
  price: Set<Price>,
  audience: Set<Audience>,
  venueDetails: Set<VenueDetail>,
  accessibilityOptions: Set<AccessibilityOption>,
  area: Set<Area>
};

export type TagFilterSet =
  | Set<Area>
  | Set<Price>
  | Set<Audience>
  | Set<VenueDetail>
  | Set<AccessibilityOption>
  | Set<Time>;

export type StringFilterSet =
  | Set<EventCategoryName>
  | Set<Audience>
  | Set<VenueDetail>
  | Set<AccessibilityOption>;

export type State = {
  selectedFilters: FilterCollection,
  stagedFilters: FilterCollection
};
