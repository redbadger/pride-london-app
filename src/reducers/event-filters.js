// @flow
import type { Reducer } from "redux";
import type { EventFiltersAction } from "../actions/event-filters";
import type { State } from "../data/event-filters";

const defaultState = {
  selectedFilters: {
    categories: new Set(), // When this is empty it signifies no category filter.
    date: null,
    timeOfDay: new Set(),
    price: new Set(),
    audience: new Set(),
    venueDetails: new Set(),
    accessibilityOptions: new Set(),
    area: new Set()
  },
  stagedFilters: {
    categories: new Set(), // When this is empty it signifies no category filter.
    date: null,
    timeOfDay: new Set(),
    price: new Set(),
    audience: new Set(),
    venueDetails: new Set(),
    accessibilityOptions: new Set(),
    area: new Set()
  }
};

const eventFilters: Reducer<State, EventFiltersAction> = (
  state: State = defaultState,
  action: EventFiltersAction
) => {
  switch (action.type) {
    case "STAGE_EVENT_FILTERS":
      return {
        ...state,
        stagedFilters: {
          ...state.stagedFilters,
          ...action.payload
        }
      };
    case "COMMIT_EVENT_FILTERS":
      return {
        ...state,
        selectedFilters: state.stagedFilters
      };
    case "CLEAR_STAGED_EVENT_FILTERS":
      return {
        ...state,
        stagedFilters: state.selectedFilters
      };
    default:
      return state;
  }
};

export default eventFilters;
