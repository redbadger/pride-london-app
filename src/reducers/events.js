// @flow
import type { Reducer } from "redux";
import type { Event } from "../integrations/cms";
import type { EventsAction } from "../actions/events";

type State = {
  events: Event[],
  loading: boolean,
  refreshing: boolean
};

const defaultState = {
  events: [],
  loading: true,
  refreshing: false
};

const events: Reducer<State, EventsAction> = (
  state: State = defaultState,
  action: EventsAction
) => {
  switch (action.type) {
    case "REQUEST_EVENTS":
      return {
        ...state,
        loading: true,
        refreshing: false
      };
    case "REQUEST_UPDATE_EVENTS":
      return {
        ...state,
        loading: false,
        refreshing: true
      };
    case "RECEIVE_EVENTS":
      return {
        ...state,
        loading: false,
        refreshing: false,
        events: action.payload ? action.payload.events : state.events
      };
    default:
      return state;
  }
};
export default events;
