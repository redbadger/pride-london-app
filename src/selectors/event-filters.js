// @flow
import {
  buildDateRangeFilter,
  buildTimeFilter,
  buildPriceFilter,
  buildStringSetFilter,
  buildAreaFilter,
  buildCategoryFilter
} from "./basic-event-filters";
import tags from "../data/tags";
import type { Event } from "../data/event";
import type { Time } from "../data/date-time";
import type { State } from "../reducers";
import type { Area } from "../data/event-filters";

const getEventFiltersState = (state: State, selectStagedFilters: boolean) =>
  selectStagedFilters
    ? state.eventFilters.stagedFilters
    : state.eventFilters.selectedFilters;

export const selectDateFilter = (
  state: State,
  selectStagedFilters?: boolean = false
) => getEventFiltersState(state, selectStagedFilters).date;
export const selectTimeFilter = (
  state: State,
  selectStagedFilters?: boolean = false
) => getEventFiltersState(state, selectStagedFilters).timeOfDay;

export const selectTagFilterSelectedCount = (
  state: State,
  selectStagedFilters?: boolean = false
) => {
  const eventFilters = getEventFiltersState(state, selectStagedFilters);
  return Object.keys(tags).reduce(
    (acc, tagName) =>
      acc + (eventFilters[tagName] ? eventFilters[tagName].size : 0),
    0
  );
};

export const selectIsStagingFilters = (state: State): boolean =>
  state.eventFilters.stagedFilters !== state.eventFilters.selectedFilters;

const buildTimesFilter = (times: Time[]) => {
  const filters = times.map(time => buildTimeFilter(time));
  return (event: Event) => filters.some(filter => filter(event));
};

const buildAreasFilter = (areas: Area[]) => {
  const filters = areas.map(area => buildAreaFilter(area));
  return (event: Event) => filters.some(filter => filter(event));
};

export const buildEventFilter = (
  state: State,
  selectStagedFilters?: boolean = false
) => {
  const {
    date,
    timeOfDay,
    price,
    audience,
    venueDetails,
    accessibilityOptions,
    area,
    categories
  } = getEventFiltersState(state, selectStagedFilters);
  const timeArray = Array.from(timeOfDay);
  const areaArray = Array.from(area);
  const dateFilter: (event: Event) => boolean = date
    ? buildDateRangeFilter(date)
    : () => true;
  const timeFilter: (event: Event) => boolean =
    timeArray.length > 0 && timeArray.length < 3
      ? buildTimesFilter(timeArray)
      : () => true;
  const priceFilter: (event: Event) => boolean = price.has("free")
    ? buildPriceFilter()
    : () => true;
  const audienceFilter: (event: Event) => boolean =
    audience.size > 0 ? buildStringSetFilter("audience", audience) : () => true;
  const venueDetailsFilter: (event: Event) => boolean =
    venueDetails.size > 0
      ? buildStringSetFilter("venueDetails", venueDetails)
      : () => true;
  const accessibilityOptionsFilter: (event: Event) => boolean =
    accessibilityOptions.size > 0
      ? buildStringSetFilter("accessibilityOptions", accessibilityOptions)
      : () => true;
  const areaFilter: (event: Event) => boolean =
    areaArray.length > 0 ? buildAreasFilter(areaArray) : () => true;
  const categoryFilter: (event: Event) => boolean =
    categories.size > 0 ? buildCategoryFilter(categories) : () => true;

  return (event: Event) =>
    dateFilter(event) &&
    timeFilter(event) &&
    priceFilter(event) &&
    audienceFilter(event) &&
    venueDetailsFilter(event) &&
    accessibilityOptionsFilter(event) &&
    areaFilter(event) &&
    categoryFilter(event);
};
