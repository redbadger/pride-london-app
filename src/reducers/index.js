// @flow
import { combineReducers } from "redux";
import events from "./events";
import type { State as EventsState } from "./events";
import eventFilters from "./event-filters";
import type { State as EventFiltersState } from "../data/event-filters";
import savedEvents from "./saved-events";
import type { SavedEvents as SavedEventsState } from "../data/event";

export type State = {
  events: EventsState,
  eventFilters: EventFiltersState,
  savedEvents: SavedEventsState
};

export default combineReducers({
  events,
  eventFilters,
  savedEvents
});
