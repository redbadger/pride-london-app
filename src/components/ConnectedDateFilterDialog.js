// @flow
import { connect } from "react-redux";
import type { Connector } from "react-redux";
import {
  stageEventFilters,
  commitEventFilters,
  clearStagedEventFilters
} from "../actions/event-filters";
import type { DateRange } from "../data/date-time";
import { selectFilteredEvents } from "../selectors/events";
import {
  selectDateFilter,
  selectIsStagingFilters
} from "../selectors/event-filters";
import Component from "./DateRangePickerDialog";
import text from "../constants/text";

type OwnProps = {
  onApply: () => void,
  onCancel: () => void,
  visible: boolean
};

type Props = {
  applyButtonText: string,
  dateRange: ?DateRange,
  onChange: (?DateRange) => void,
  forceNewRange: boolean
} & OwnProps;

const mapStateToProps = state => ({
  applyButtonText: text.filterPickerApply(
    selectFilteredEvents(state, true).length
  ),
  dateRange: selectDateFilter(state, true),
  forceNewRange: !selectIsStagingFilters(state)
});

const mapDispatchToProps = (dispatch, ownProps) => ({
  onChange: date => dispatch(stageEventFilters({ date })),
  onApply: () => {
    ownProps.onApply();
    dispatch(commitEventFilters());
  },
  onCancel: () => {
    ownProps.onCancel();
    dispatch(clearStagedEventFilters());
  }
});

const connector: Connector<OwnProps, Props> = connect(
  mapStateToProps,
  mapDispatchToProps
);

export default connector(Component);
