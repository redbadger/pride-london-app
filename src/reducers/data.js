// @flow
import type { Reducer } from "redux";
import type { DataAction } from "../actions/data";
import type { CmsEntry } from "../integrations/cms";
import type { Asset } from "../data/asset";
import { expandRecurringEventsInEntries } from "../selectors/events";

export type State = {
  entries: CmsEntry[],
  assets: Asset[],
  loading: boolean,
  refreshing: boolean
};

const defaultState = {
  entries: [],
  assets: [],
  loading: true,
  refreshing: false
};

const processEntries = entries => expandRecurringEventsInEntries(entries);

const reducer: Reducer<State, DataAction> = (
  state: State = defaultState,
  action: DataAction
) => {
  switch (action.type) {
    case "REQUEST_CMS_DATA":
      return {
        ...state,
        loading: true,
        refreshing: false
      };
    case "REQUEST_UPDATE_CMS_DATA":
      return {
        ...state,
        loading: false,
        refreshing: true
      };
    case "RECEIVE_CMS_DATA":
      return {
        loading: false,
        refreshing: false,
        entries: processEntries(action.data.entries),
        assets: action.data.assets
      };
    default:
      return state;
  }
};
export default reducer;
