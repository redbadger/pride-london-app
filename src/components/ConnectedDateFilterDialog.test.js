import React from "react";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import { shallow } from "enzyme";
import { DateTime } from "luxon";
import ConnectedDateFilterDialog from "./ConnectedDateFilterDialog";
import { createEventFiltersState } from "../reducers/event-filters";

const initialState = {
  data: {
    entries: [],
    assets: [],
    loading: true,
    refreshing: false
  },
  eventFilters: createEventFiltersState(
    DateTime.fromISO("2018-07-07T00:00:00+01:00")
  )
};

const singleDateFilters = {
  categories: new Set(),
  date: { startDate: "2018-02-02", endDate: "2018-02-02" },
  timeOfDay: new Set(),
  price: new Set(),
  audience: new Set(),
  venueDetails: new Set(),
  accessibilityOptions: new Set(),
  area: new Set()
};

const singleDateFilterState = {
  ...initialState,
  eventFilters: {
    selectedFilters: singleDateFilters,
    stagedFilters: singleDateFilters
  }
};

const mockStore = configureStore([thunk]);

describe("ConnectedDateFilterDialog", () => {
  it("renders connector", () => {
    const store = mockStore(initialState);
    const output = shallow(
      <ConnectedDateFilterDialog
        store={store}
        onApply={() => {}}
        onCancel={() => {}}
        visible
      />
    );
    expect(output).toMatchSnapshot();
  });

  it("renders component", () => {
    const store = mockStore(initialState);
    const output = shallow(
      <ConnectedDateFilterDialog
        store={store}
        onApply={() => {}}
        onCancel={() => {}}
        visible
      />
    );
    expect(output.dive()).toMatchSnapshot();
  });

  it("commits filters on apply", () => {
    const store = mockStore(initialState);
    const mockOnApply = jest.fn();
    const output = shallow(
      <ConnectedDateFilterDialog
        store={store}
        onApply={mockOnApply}
        onCancel={() => {}}
        visible
      />
    );

    output
      .dive()
      .props()
      .onApply();
    const actions = store.getActions();

    expect(mockOnApply).toHaveBeenCalled();
    expect(actions).toEqual([{ type: "COMMIT_EVENT_FILTERS" }]);
  });

  it("clears filters on cancel", () => {
    const store = mockStore(initialState);
    const mockOnCancel = jest.fn();
    const output = shallow(
      <ConnectedDateFilterDialog
        store={store}
        onApply={() => {}}
        onCancel={mockOnCancel}
        visible
      />
    );

    output
      .dive()
      .props()
      .onCancel();
    const actions = store.getActions();

    expect(mockOnCancel).toHaveBeenCalled();
    expect(actions).toEqual([{ type: "CLEAR_STAGED_EVENT_FILTERS" }]);
  });

  it("updates filters on change", () => {
    const store = mockStore(initialState);
    const output = shallow(
      <ConnectedDateFilterDialog
        store={store}
        onApply={() => {}}
        onCancel={() => {}}
        visible
      />
    );

    output
      .dive()
      .find("DateRangePicker")
      .props()
      .onChange("2018-02-02");

    const actions = store.getActions();

    expect(actions).toEqual([
      { type: "STAGE_EVENT_FILTERS", payload: { date: "2018-02-02" } }
    ]);
  });

  it("forces a new range when not staging filters", () => {
    const store = mockStore(singleDateFilterState);
    const output = shallow(
      <ConnectedDateFilterDialog
        store={store}
        onApply={() => {}}
        onCancel={() => {}}
        visible
      />
    );

    expect(output.prop("forceNewRange")).toBe(true);
  });
});
