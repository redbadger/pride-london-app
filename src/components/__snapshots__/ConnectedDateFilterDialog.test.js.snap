// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConnectedDateFilterDialog renders component 1`] = `
<Dialog
  applyButtonLabel="Show events (none selected)"
  applyButtonText="No events"
  headerRight={
    <Touchable
      accessibilityComponentType="button"
      accessibilityLabel="Clear date selection"
      accessibilityTraits={
        Array [
          "button",
        ]
      }
      accessible={undefined}
      delayLongPress={undefined}
      delayPressIn={50}
      delayPressOut={undefined}
      disabled={undefined}
      hitSlop={undefined}
      onLongPress={undefined}
      onPress={[Function]}
      onPressIn={undefined}
      onPressOut={undefined}
      pressRetentionOffset={undefined}
      style={undefined}
    >
      <Text
        color="blackColor"
        markdown={false}
        markdownStyle={Object {}}
        style={
          Object {
            "color": "rgb(45, 47, 127)",
          }
        }
        type="small"
      >
        Clear
      </Text>
    </Touchable>
  }
  onApply={[Function]}
  onCancel={[Function]}
  title={
    <Text
      accessibilityLabel="Select dates"
      accessibilityLiveRegion="polite"
      accessibilityTraits={
        Array [
          "header",
        ]
      }
      accessible={true}
      color="blackColor"
      markdown={false}
      markdownStyle={Object {}}
      style={
        Object {
          "alignSelf": "flex-end",
          "color": "rgb(45, 47, 127)",
          "height": 40,
          "paddingTop": 10,
        }
      }
      type="h3"
    >
      Select dates
    </Text>
  }
  visible={true}
>
  <DateRangePicker
    dateRange={null}
    forceNewRange={false}
    onChange={[Function]}
  />
</Dialog>
`;

exports[`ConnectedDateFilterDialog renders connector 1`] = `
<DateRangePickerDialog
  applyButtonLabel="Show events (none selected)"
  applyButtonText="No events"
  dateRange={null}
  forceNewRange={false}
  onApply={[Function]}
  onCancel={[Function]}
  onChange={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
  visible={true}
/>
`;
